fix(compiler): changed after checked error in for loops

crisbeto opened the pull request:
Reworks the `repeater` instruction to go through `advance`, instead of passing in the index directly. This ensures that lifecycle hooks run at the right time and that we don't throw "changed after checked" errors when we shouldn't be.

Fixes #52885.

pkozlowski-opensource commented on packages/core/test/acceptance/control_flow_for_spec.ts:
```diff
@@ -121,6 +121,41 @@ describe('control flow - for', () => {
+    expect(() => {
+      const fixture = TestBed.createComponent(TestComponent);
+      fixture.detectChanges();
+    }).not.toThrow();
```
could we actually have an assert here verifying the actual output?

pkozlowski-opensource replied:
Thnx. Now you can remove the expect for not throwing (this would break the test automatically, right?)

crisbeto replied:
Done

crisbeto committed:
fix(compiler): changed after checked error in for loops

Reworks the `repeater` instruction to go through `advance`, instead of passing in the index directly. This ensures that lifecycle hooks run at the right time and that we don't throw "changed after checked" errors when we shouldn't be.

Fixes #52885.

jessicajaniuk commented:
This PR was merged into the repository by commit ec2d6e7b9c2b386247d1320ee89f8e3ac5e5a0dd.

dylhunn reviewed (commented):
Thank you for fixing template pipeline :)
